lib/Backend/Commands.hs

```diff
@@ -329,6 +339,70 @@ copyCmd (CopyOptions oldPath newPath force) = do
     getOperationPaths :: CopyOperation -> (EntryPath, EntryPath)
     getOperationPaths (CopyOperation old new) = (old ^. E.path, new ^. E.path)
 
+    checkForEntriesInPath :: [CopyOperation] -> Sem r ()
+    checkForEntriesInPath operations = do
+      invalidOperations <- flip filterM operations \(CopyOperation _ new) -> do
+        let path = Path.entryPathAsPath $ new ^. E.path
+        if length (path ^. pathSegments) > 1 then do
+          let pathInits = Path.EntryPath . NE.fromList <$> (init . tail . inits) (path ^. pathSegments)
+          filterM pathIsEntry pathInits >>= \case
+            [] -> pure False
+            _  -> pure True
+        else
+          pure False
+      when (not (null invalidOperations)) do
+        invalidOperations
+          <&> (\(CopyOperation old new) -> (old ^. E.path, new ^. E.path) )
+          & throw . CPRParentDirectoryIsEntry
+
+    checkForEntriesInSubpaths :: [CopyOperation] -> Sem r ()
+    checkForEntriesInSubpaths operations = do
+      invalidOperations <- flip filterM operations \(CopyOperation old new) -> do
+        checkForEntriesInSubpaths_ (Path.entryPathAsPath $ old ^. E.path) (Path.entryPathAsPath $ new ^. E.path)
+      when (not (null invalidOperations)) do
+        invalidOperations
+          <&> (\(CopyOperation old new) -> (old ^. E.path, new ^. E.path) )
+          & throw . CPRSubdirectoriesContainEntries
+      where
+        checkForEntriesInSubpaths_ :: Path -> Path -> Sem r Bool
+        checkForEntriesInSubpaths_ path1 path2 = do
+          mDir1 <- getEntryOrDir path1
+          mDir2 <- getEntryOrDir path2
+          if _Just `is` mDir1 && _Just `is` mDir2 then do
+            let eDir1 = mDir1 ^?! _Just
+                eDir2 = mDir2 ^?! _Just
+            if _Right `is` eDir1 && _Right `is` eDir2 then do
+              let dir1 = eDir1 ^?! _Right
+                  dir2 = eDir2 ^?! _Right
+              checkForEntriesInSubdirs dir1 dir2
+            else
+              pure False
+          else
+            pure False
+
+        checkForEntriesInSubdirs :: Directory -> Directory -> Sem r Bool
+        checkForEntriesInSubdirs dir1 dir2 = do
+          -- 'subdirs' hashmap at first level contains start directory
+          entries <- execStateT (getEqualDirectoriesAndEntries (head . HashMap.elems $ dir1 ^. Dir.subdirs) (head . HashMap.elems $ dir2 ^. Dir.subdirs)) []
```

Comment:
`head` is an unsafe function, can we avoid using it?

Reply:
Done